### PR TITLE
[Bug] Fix #1340

### LINKED ISSUE #1340: [BUG] RedisEmbeddingStore returns embeddings from different indexes
**Describe the bug**
During  search RedisEmbeddingStore finds embeddings in all existing Redis embedding stores, in the same instance of Redis (not only current one) .
Bug was noticed around version 0.23.0, when TODO comment was added in the RedisEmbeddingStoreIT:
```
// TODO fix: why redis returns embeddings from different indexes?
```

**To Reproduce**
Create at least two Redis storages using the same Redis instance.
Embeddings added to the first storage would be found when search is done with use of second storage.

It's also possible to run this test (can be added to RedisEmbeddingStoreIT):
 ```	
    @Test
    void should_find_embedding_in_proper_store() {
       final EmbeddingStore<TextSegment> firstStore = createStore("first-store");
       final EmbeddingStore<TextSegment> secondStore = createStore("second-store");

       Embedding embedding = embeddingModel().embed("hello").content();
       
       String id = firstStore.add(embedding);
       assertThat(id).isNotBlank();

       List<EmbeddingMatch<TextSegment>> relevant = firstStore.findRelevant(embedding, 10);
       assertThat(relevant).hasSize(1);
       assertThat(secondStore.findRelevant(embedding, 10)).hasSize(0);

       // new API
       assertThat(firstStore.search(EmbeddingSearchRequest.builder()
               .queryEmbedding(embedding)
               .maxResults(10)
               .build()).matches()).isEqualTo(relevant);
       assertThat(secondStore.search(EmbeddingSearchRequest.builder()
               .queryEmbedding(embedding)
               .maxResults(10)
               .build()).matches()).hasSize(0);
    }
    
    private EmbeddingStore<TextSegment> createStore(String name) {
      return RedisEmbeddingStore.builder()
          .host(redis.getHost())
          .port(redis.getFirstMappedPort())
          .indexName(name)
          .dimension(384)
          .metadataKeys(createMetadata().toMap().keySet())
          .build();      
  }
```    

Current tests in the RedisEmbeddingStoreIT  can also be used to reproduce, when lines:
```
        try (JedisPooled jedis = new JedisPooled(redis.getHost(), redis.getFirstMappedPort())) {
            jedis.flushDB(); // TODO fix: why redis returns embeddings from different indexes?
        }
```
are removed.

**Expected behavior**
RedisEmbeddingStore should only return embeddings from current storage

**Please complete the following information:**
- LangChain4j version: 0.31.0 
- LLM(s) used: llama3
- Java version: openjdk 21.0.2 2024-01-16 LTS
- Spring Boot version (if applicable): 3.2.5

**Additional context**
Problem is caused by having the same key prefix (defined in RedisSchema as "embedding:") for all embeddings stored in the Redis.
Since that all Redis indexes created by langchain4j are searching through the same embedding source.
(I'm currently working on PR)

### PR DESCRIPTION
**Pull request by Martin7-1**
## Issue

Closes #1340 

## Change

1. Make `prefix` configurable and default value is backwards compatible (i.e. the default value is `embedding:`).
2. Remove `RedisEmbeddingStoreIT` TODO.

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

## Checklist for changing existing embedding store integration

- [x] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Review by kpavlov — APPROVED** [review]
Thank you, @Martin7-1

**Review by Wadimz — APPROVED** [review]

**Review by Wadimz — APPROVED** [review]

**Comment by Martin7-1** [comment]
@langchain4j Hi! This PR is ready to merge. Github Actions failed because of the unit tests failed, but it's not related to this PR :). @kpavlov Could you please check the unit tests?

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java [hunk 1]
```diff
@@ -34,7 +34,17 @@ class RedisSchema {
...
+    RedisSchema(String indexName,
+                String prefix,
+                String vectorFieldName,
+                String scalarFieldName,
+                Collection<String> metadataKeys,
+                VectorAlgorithm vectorAlgorithm,
+                int dimension,
+                MetricType metricType) {
+        if (!prefix.endsWith(":")) {
+            prefix += ":";
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java]: Should we do this? Documentation says:

> should be ended with ":" (e.g. embedding:)

So I would either remove or document this behavior

**Review by ghost — COMMENTED** [review]
@Martin7-1 thank you!

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java (hunk 1 above)
Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java]: Sounds reasonable. Currently I remove it and replace with check `prefix.endWith(":")` is true.

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 2]
```diff
@@ -276,6 +290,15 @@ public Builder indexName(String indexName) {
             return this;
         }
 
+        /**
+         * @param prefix The prefix of the document, should be ended with ":" (e.g. embedding:)
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]:
```suggestion
         * @param prefix The prefix of the key, should end with a colon (e.g., "embedding:") (optional). Default value: "embedding:".
```

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 3]
```diff
@@ -48,6 +60,7 @@ public class RedisEmbeddingStore implements EmbeddingStore<TextSegment> {
      * @param user         Redis Stack username (optional)
      * @param password     Redis Stack password (optional)
      * @param indexName    The name of the index (optional). Default value: "embedding-index".
+     * @param prefix       The prefix of the document, should be ended with ":" (e.g. embedding:) (optional). Default value: "embedding:"
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]:
```suggestion
     * @param prefix       The prefix of the key, should end with a colon (e.g., "embedding:") (optional). Default value: "embedding:".
```

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java [hunk 4]
```diff
@@ -34,7 +39,16 @@ class RedisSchema {
...
-    RedisSchema(String indexName, String prefix, String vectorFieldName, String scalarFieldName, Collection<String> metadataKeys, VectorAlgorithm vectorAlgorithm, int dimension, MetricType metricType) {
+    RedisSchema(String indexName,
+                String prefix,
+                String vectorFieldName,
+                String scalarFieldName,
+                Collection<String> metadataKeys,
+                VectorAlgorithm vectorAlgorithm,
+                int dimension,
+                MetricType metricType) {
+        ensureTrue(prefix.endsWith(":"), "Prefix should be ended with ':'");
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java]:
```suggestion
        ensureTrue(prefix.endsWith(":"), "Prefix should end with a ':'");
```

**Review by ghost — DISMISSED** [review]
@Martin7-1 thank you!

**Pull request merged**